### PR TITLE
8329961: Buffer overflow in os::Linux::kernel_version

### PR DESCRIPTION
**Pull request by jdksjolen**
Hi,

There was a bug in the original implementation of `os::Linux::kernel_version` which this PR fixes. Namely, the comparison `walker != nullptr` is wrong, the intended comparison was `*walker != '\0'` or `walker[0] != '\0'`. This means that if a bad/unexpected version string is encountered the `walker` would read past the string.

We fix this by applying the correct comparison and adding some basic tests.

@luhenry , @robehn. You attempted to create automatic backport branches on this in the original PR, can you check whether this fix also needs to be backported to the mentioned versions? The original PR link is this: https://github.com/openjdk/jdk/pull/17889

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8329961](https://bugs.openjdk.org/browse/JDK-8329961): Buffer overflow in os::Linux::kernel_version (**Bug** - P4)

### Reviewers
 * [Robbin Ehn](https://openjdk.org/census#rehn) (@robehn - **Reviewer**)
 * [Thomas Stuefe](https://openjdk.org/census#stuefe) (@tstuefe - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/18697/head:pull/18697` \
`$ git checkout pull/18697`

Update a local copy of the PR: \
`$ git checkout pull/18697` \
`$ git pull https://git.openjdk.org/jdk.git pull/18697/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 18697`

View PR using the GUI difftool: \
`$ git pr show -t 18697`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/18697.diff">https://git.openjdk.org/jdk/pull/18697.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/18697#issuecomment-2045227571)

### CODE REVIEW

**Review by robehn — APPROVED** [review]
Thanks you!

Yes it should be backported to jdk21u-dev and jdk22u.

File: src/hotspot/os/linux/os_linux.cpp [hunk 1]
```diff
@@ -367,6 +367,19 @@ static void next_line(FILE *f) {
   } while (c != '\n' && c != EOF);
 }
 
+void os::Linux::parse_kernel_version(long* major, long* minor, char* release) {
```
Comment (tstuefe) [inline src/hotspot/os/linux/os_linux.cpp:370]: release should be const char*, and possibly at the start.

**Review by tstuefe — COMMENTED** [review]
Hmm, wouldn't sscanf not be simpler and safer? No need to factor out the parser. IMHO no need to even add a gtest since parsing would be really simple and not loop based. E.g.

```
if (sscanf(release, "%d.%d", &major, &minor) != 2) {
  log_warning blabla
}
```

As bonus, you avoid accidental conversion from hex numbers and such that strotol provides and that we don't really want here.

**Comment by tstuefe** [comment]
> As bonus, you avoid accidental conversion from hex numbers and such that strotol provides and that we don't really want here.

Okay scratch the last sentence, since you manually specify the base, this would not be an issue.

**Comment by robehn** [comment]
> Another thing: We shouldn't call `uname`. Reading `/proc/sys/kernel/osrelease` is sufficient.

I don't follow that. uname() is POSIX portable, while /proc/sys/kernel/osrelease is Linux specific.
Now this is code is in Linux part, but I don't see why a less portable way that require more code would be superior?

**Comment by jdksjolen** [comment]
> > > Hmm, wouldn't sscanf not be simpler and safer? No need to factor out the parser. IMHO no need to even add a gtest since parsing would be really simple and not loop based. E.g.
> > > ```
> > > if (sscanf(release, "%d.%d", &major, &minor) != 2) {
> > >   log_warning blabla
> > > }
> > > ```
> > > 
> > > 
> > >     
> > >       
> > >     
> > > 
> > >       
> > >     
> > > 
> > >     
> > >   
> > > As bonus, you avoid accidental conversion from hex numbers and such that strotol provides and that we don't really want here.
> > 
> > 
> > Hi, according to C11 standard (and my man pages) it is UB to call the scanf-family of functions with "invalid" data and `strtol` is recommended instead. So, unfortunately, it might not be safer.
> 
> Can you point me to the relevant man page? If that were not to work, it would drastically limit the usefulness of scanf, and a lot of code in hotspot would be UB.
> 
> My manpage:
> 
> ```
> The format string consists of a sequence of directives which describe how to process the sequence of input characters. If processing of a directive fails, no further input is read, and scanf() returns. A "failure" can be either of the following: input failure, meaning that input characters were unavailable, or matching failure, meaning that the input was inappropriate (see below). 
> ```
> 
> Matching failures are part of the API.
> 
> > Another thing: We shouldn't call `uname`. Reading `/proc/sys/kernel/osrelease` is sufficient.
> 
> I agree with @robehn, uname is Posix and portable. May also be cached inside glibc, saving us a proc fs read.

See BUGS section: https://man7.org/linux/man-pages/man3/sscanf.3.html


OK, I'm probably too paranoid regarding `uname`. I don't like that we've got a C string and there's no way of knowing whether it actually ends in a NUL byte or not. At least with a file you know that it has a specific length and can account for that. We probably have to assume that if someone is being malicious with `uname` then we're compromised regardless.

**Comment by tstuefe** [comment]
> > > > Hmm, wouldn't sscanf not be simpler and safer? No need to factor out the parser. IMHO no need to even add a gtest since parsing would be really simple and not loop based. E.g.
> > > > ```
> > > > if (sscanf(release, "%d.%d", &major, &minor) != 2) {
> > > >   log_warning blabla
> > > > }
> > > > ```
> > > > 
> > > > 
> > > >     
> > > >       
> > > >     
> > > > 
> > > >       
> > > >     
> > > > 
> > > >     
> > > >   
> > > > As bonus, you avoid accidental conversion from hex numbers and such that strotol provides and that we don't really want here.
> > > 
> > > 
> > > Hi, according to C11 standard (and my man pages) it is UB to call the scanf-family of functions with "invalid" data and `strtol` is recommended instead. So, unfortunately, it might not be safer.
> > 
> > 
> > Can you point me to the relevant man page? If that were not to work, it would drastically limit the usefulness of scanf, and a lot of code in hotspot would be UB.
> > My manpage:
> > ```
> > The format string consists of a sequence of directives which describe how to process the sequence of input characters. If processing of a directive fails, no further input is read, and scanf() returns. A "failure" can be either of the following: input failure, meaning that input characters were unavailable, or matching failure, meaning that the input was inappropriate (see below). 
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > Matching failures are part of the API.
> > > Another thing: We shouldn't call `uname`. Reading `/proc/sys/kernel/osrelease` is sufficient.
> > 
> > 
> > I agree with @robehn, uname is Posix and portable. May also be cached inside glibc, saving us a proc fs read.
> 
> See BUGS section: https://man7.org/linux/man-pages/man3/sscanf.3.html
> 

Okay that. Note that deprecation of numerical conversions has been hugely controversial, and it was a step taken by glibc manpage maintainers. This is just about numerical overflow. E.g. 3333333333333333333333333333333.333333333333333333333333333333 will overflow int range, and return true. Are we honestly concerned about this here?

By avoiding sscanf, we swap very readable and simple-hence-safe code with manual parsing, which is both unreadable and, as this issue shows, in practice a lot unsafer than had we used sscanf to begin with.

> OK, I'm probably too paranoid regarding `uname`. I don't like that we've got a C string and there's no way of knowing whether it actually ends in a NUL byte or not. At least with a file you know that it has a specific length and can account for that. We probably have to assume that if someone is being malicious with `uname` then we're compromised regardless.

If you are paranoid, do this:

```
struct {
  utsname uts; char c;
} v;
```

then set v.c to 0 before calling uname.

**Comment by jdksjolen** [comment]
> > > > > Hmm, wouldn't sscanf not be simpler and safer? No need to factor out the parser. IMHO no need to even add a gtest since parsing would be really simple and not loop based. E.g.
> > > > > ```
> > > > > if (sscanf(release, "%d.%d", &major, &minor) != 2) {
> > > > >   log_warning blabla
> > > > > }
> > > > > ```
> > > > > 
> > > > > 
> > > > >     
> > > > >       
> > > > >     
> > > > > 
> > > > >       
> > > > >     
> > > > > 
> > > > >     
> > > > >   
> > > > > As bonus, you avoid accidental conversion from hex numbers and such that strotol provides and that we don't really want here.
> > > > 
> > > > 
> > > > Hi, according to C11 standard (and my man pages) it is UB to call the scanf-family of functions with "invalid" data and `strtol` is recommended instead. So, unfortunately, it might not be safer.
> > > 
> > > 
> > > Can you point me to the relevant man page? If that were not to work, it would drastically limit the usefulness of scanf, and a lot of code in hotspot would be UB.
> > > My manpage:
> > > ```
> > > The format string consists of a sequence of directives which describe how to process the sequence of input characters. If processing of a directive fails, no further input is read, and scanf() returns. A "failure" can be either of the following: input failure, meaning that input characters were unavailable, or matching failure, meaning that the input was inappropriate (see below). 
> > > ```
> > > 
> > > 
> > >     
> > >       
> > >     
> > > 
> > >       
> > >     
> > > 
> > >     
> > >   
> > > Matching failures are part of the API.
> > > > Another thing: We shouldn't call `uname`. Reading `/proc/sys/kernel/osrelease` is sufficient.
> > > 
> > > 
> > > I agree with @robehn, uname is Posix and portable. May also be cached inside glibc, saving us a proc fs read.
> > 
> > 
> > See BUGS section: https://man7.org/linux/man-pages/man3/sscanf.3.html
> 
> Okay that. Note that deprecation of numerical conversions has been hugely controversial, and it was a step taken by glibc manpage maintainers. This is just about numerical overflow. E.g. 3333333333333333333333333333333.333333333333333333333333333333 will overflow int range, and return true. Are we honestly concerned about this here?
> 
> By avoiding sscanf, we swap very readable and simple-hence-safe code with manual parsing, which is both unreadable and, as this issue shows, in practice a lot unsafer than had we used sscanf to begin with.
> 
> > OK, I'm probably too paranoid regarding `uname`. I don't like that we've got a C string and there's no way of knowing whether it actually ends in a NUL byte or not. At least with a file you know that it has a specific length and can account for that. We probably have to assume that if someone is being malicious with `uname` then we're compromised regardless.
> 
> If you are paranoid, do this:
> 
> ```
> struct {
>   utsname uts; char c;
> } v;
> ```
> 
> then set v.c to 0 before calling uname.

Aha, it's just overflow issues? Well, then let's just use `sscanf`. Thanks for the hint, interesting code pattern! That won't work as the string isn't necessarily embedded within the struct, but it's an interesting strategy nonetheless. I'll just trust that it ends at some point.

**Comment by tstuefe** [comment]
> > > > > > Hmm, wouldn't sscanf not be simpler and safer? No need to factor out the parser. IMHO no need to even add a gtest since parsing would be really simple and not loop based. E.g.
> > > > > > ```
> > > > > > if (sscanf(release, "%d.%d", &major, &minor) != 2) {
> > > > > >   log_warning blabla
> > > > > > }
> > > > > > ```
> > > > > > 
> > > > > > 
> > > > > >     
> > > > > >       
> > > > > >     
> > > > > > 
> > > > > >       
> > > > > >     
> > > > > > 
> > > > > >     
> > > > > >   
> > > > > > As bonus, you avoid accidental conversion from hex numbers and such that strotol provides and that we don't really want here.
> > > > > 
> > > > > 
> > > > > Hi, according to C11 standard (and my man pages) it is UB to call the scanf-family of functions with "invalid" data and `strtol` is recommended instead. So, unfortunately, it might not be safer.
> > > > 
> > > > 
> > > > Can you point me to the relevant man page? If that were not to work, it would drastically limit the usefulness of scanf, and a lot of code in hotspot would be UB.
> > > > My manpage:
> > > > ```
> > > > The format string consists of a sequence of directives which describe how to process the sequence of input characters. If processing of a directive fails, no further input is read, and scanf() returns. A "failure" can be either of the following: input failure, meaning that input characters were unavailable, or matching failure, meaning that the input was inappropriate (see below). 
> > > > ```
> > > > 
> > > > 
> > > >     
> > > >       
> > > >     
> > > > 
> > > >       
> > > >     
> > > > 
> > > >     
> > > >   
> > > > Matching failures are part of the API.
> > > > > Another thing: We shouldn't call `uname`. Reading `/proc/sys/kernel/osrelease` is sufficient.
> > > > 
> > > > 
> > > > I agree with @robehn, uname is Posix and portable. May also be cached inside glibc, saving us a proc fs read.
> > > 
> > > 
> > > See BUGS section: https://man7.org/linux/man-pages/man3/sscanf.3.html
> > 
> > 
> > Okay that. Note that deprecation of numerical conversions has been hugely controversial, and it was a step taken by glibc manpage maintainers. This is just about numerical overflow. E.g. 3333333333333333333333333333333.333333333333333333333333333333 will overflow int range, and return true. Are we honestly concerned about this here?
> > By avoiding sscanf, we swap very readable and simple-hence-safe code with manual parsing, which is both unreadable and, as this issue shows, in practice a lot unsafer than had we used sscanf to begin with.
> > > OK, I'm probably too paranoid regarding `uname`. I don't like that we've got a C string and there's no way of knowing whether it actually ends in a NUL byte or not. At least with a file you know that it has a specific length and can account for that. We probably have to assume that if someone is being malicious with `uname` then we're compromised regardless.
> > 
> > 
> > If you are paranoid, do this:
> > ```
> > struct {
> >   utsname uts; char c;
> > } v;
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > then set v.c to 0 before calling uname.
> 
> Aha, it's just overflow issues? Well, then let's just use `sscanf`. Thanks for the hint, interesting code pattern! That won't work as the string isn't necessarily embedded within the struct, but it's an interesting strategy nonetheless.

I think it is required to be inline, though. See https://pubs.opengroup.org/onlinepubs/009695399/basedefs/sys/utsname.h.html , definition is to be an array, not a pointer:

```
char  release[]  Current release level of this implementation. 
```

> I'll just trust that it ends at some point.

It has to. Address space is ending at 128TB usually. :-)

**Review by tstuefe — APPROVED** [review]
neat

**Comment by jdksjolen** [comment]
Cool! Thank you for the reviews.

/integrate

**Comment by robehn** [comment]
@jdksjolen are you backporting this?

**Comment by jdksjolen** [comment]
> @jdksjolen are you backporting this?

@robehn, thanks. I forgot about doing that. Good for me to have a try at doing that, but I'll need sponsorship. 

/backport jdk21u-dev
/backport jdk22u

**Pull request closed without merging**